### DISCUSSION TITLE
Field errors to bending magnets

### DISCUSSION THREAD

**Discussion by elafmusa** [atcollab/at#680] [Python AT]
Hi everyone,

I have a bit naive question,
What is the best way to assign relative field errors to a bending magnet in PyAT? I checked the arguments of the dipole elements:

> family_name: Name of the element
> length: Element length [m]
> bending_angle: Bending angle [rd]
> poly_a: Array of normal multipole components
> poly_b: Array of skew multipole components
> k=0: Field index

Is doing as following correct?

ring[dipole_ind].BendingAngle *=(1+ error_value * length)

Thank you

**Comment by simoneliuzzo** [comment]
Dear @elafmusa,

bending angles are in fact not trivial. They define the lattice geometry and the dispersion. 

1) Let's imagine what happens for a change of bending angle: 
- BendingAngle is modified to T + dt
- There is a new reference trajectory (potentially not closed)
- Dispersion changes

Instead what I do, is that I assume that BendingAngles are simply there to specify my reference. 

2) All modifications are applied to PolynomB/A. 
- PolynomB is modified to PolynomB[0] += PolynomB[0] + dt/L
- There is no change in the reference trajectory.
- Dispersion changes (differently from in the above case)
- there is a closed orbit change

The first way is the way to go in case of real lattice modifications, where the magnets will have a different field by design.
The second is in my opinion the most clear when we want to study the effect of errors. 
 
May be @lfarv and @swhite2401 have better options?

**Comment by lfarv** [comment]
Hello @elafmusa 

Relative field errors can be easily applied using the `FieldScaling` attribute. Applying it on a bending magnet gives a correct treatment of the closed orbit distortion, modification of the magnet focusing, if any, modification of pole face angle effects and distortion of the dispersion. As pointed out by @simoneliuzzo, this does not change the reference trajectory, but it generates a closed orbit distortion. But unlike @simoneliuzzo's solution, it correctly handles focusing effects.

The default value of `FieldScaling` is 1.0, corresponding to the nominal field. Setting `dipole.FieldScaling = 1.001`, for instance, corresponds to a +1.0e-3 relative field error.

By the way, the `FieldScaling` attribute may be used on any magnet (quadrupole, sextupole…). For straight magnets, it's strictly equivalent to a multiplication of the full `PolynomA` and `PolynomB` by the same amount. But it's simpler to apply or remove. For curved magnets, there is no other equivalent solution, @simoneliuzzo's proposal being the closest approximation.

Valid for both python and Matlab !

**Comment by ZeusMarti** [comment]
Hi @lfarv ,

Could you elaborate on why for curved magnets FieldScaling is not equivalent to manually scale PolynomA and PolynomB plus adding the extra term in PolynomB[0] that Simone mentioned? Apart from modifying edge effects what else is missing?

Thanks,
Zeus

**Comment by lfarv** [comment]
Hello @ZeusMarti 

A pure sector dipole (`PolynomB==0`) still has an intrinsic (geometric) focusing, with a strength 1/&rho;<sup>2</sup>. No trick on `PolynomB` can scale that. The only way is to scale &rho;, acting on `BendingAngle` or `Length`. But as @simoneliuzzo mentions, that would change the reference trajectory, which we do not want.

Practically, for large bending radii, this intrinsic focusing is small, so a small error on it is probably negligible. But `FieldScaling` gives an easy way to avoid the problem.

The way `FieldScaling` works is based on the fact that scaling a magnet field is equivalent to scaling the design momentum of the line. On entering a magnet, `FieldScaling` scales the design momentum by the same amount, and consequently modifies &delta;, p<sub>x</sub> and p<sub>y</sub> to refer to the new p<sub>0</sub>. Then it crosses the magnet and returns back to the normal momentum. Note that this does not change the particle momentum, but the reference momentum p<sub>0</sub>.

On a straight magnet, I checked that numerically it gives **exactly** the same result as scaling `PolynomA` and `PolynomB`. However, for a curved magnet, there is a small difference.

**Comment by ZeusMarti** [comment]
Thanks @lfarv ! Changing the momentum looks like a nice trick! Is this AT exclusive or inherited from other codes? In case is only AT, let me congratulate and thank you again!

**Comment by lfarv** [comment]
@ZeusMarti:

> Is this AT exclusive or inherited from other codes?


I'm not aware of anything similar, but that does not mean it's not used in some code…
